File: VB/CommanChannelServiceDataStore/CommandChannelServiceDataStore.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.Xpo.DB
Imports System.Web.Services
Imports DevExpress.Xpo.Helpers
Imports System.Web.Services.Protocols
Imports System.Xml.Serialization
Imports DevExpress.Data.Filtering
Imports System.Collections.Generic

Namespace DXSample.DataStore
	<WebServiceBinding(Namespace:="http://devexpress.example/")> _
	Public Class CommandChannelServiceDataStore
		Inherits SoapHttpClientProtocol
		Implements IDataStore, ICommandChannel
		Public Sub New(ByVal url As String)
			MyBase.New()
			Me.Url = url
		End Sub

		Protected Overridable ReadOnly Property AutoCreateOption() As AutoCreateOption
			Get
				Return AutoCreateOption.SchemaAlreadyExists
			End Get
		End Property

		<SoapDocumentMethod("http://devexpress.example/ModifyData", RequestNamespace := "http://devexpress.example/", ResponseNamespace := "http://devexpress.example/")> _
		Public Overridable Function ModifyData(ParamArray ByVal dmlStatements() As ModificationStatement) As ModificationResult
			Return CType(Invoke("ModifyData", New Object() { dmlStatements })(0), ModificationResult)
		End Function

		<SoapDocumentMethod("http://devexpress.example/SelectData", RequestNamespace := "http://devexpress.example/", ResponseNamespace := "http://devexpress.example/")> _
		Public Overridable Function SelectData(ParamArray ByVal selects() As SelectStatement) As SelectedData
			Return CType(Invoke("SelectData", New Object() { selects })(0), SelectedData)
		End Function

		Protected Overridable Function UpdateSchema(ByVal dontCreateIfFirstTableNotExist As Boolean, ParamArray ByVal tables() As DBTable) As UpdateSchemaResult
			Throw New InvalidOperationException("Schema modification is not allowed")
		End Function

		<SoapDocumentMethod("http://devexpress.example/Do", RequestNamespace := "http://devexpress.example/", ResponseNamespace := "http://devexpress.example/"), XmlInclude(GetType(StoredProcedureArgument))> _
		Public Overridable Function [Do](ByVal command As String, ByVal args As Object) As Object
			Return Invoke("Do", New Object() { command, PrepareArguments(args) })(0)
		End Function

		Private Function PrepareArguments(ByVal args As Object) As Object
			Dim arguments() As Object = TryCast(args, Object())
			If arguments Is Nothing Then
				Return args
			End If
			Return New StoredProcedureArgument(arguments(0), arguments(1))
		End Function

		#Region "IDataStore Members"

		Private ReadOnly Property IDataStore_AutoCreateOption() As AutoCreateOption Implements IDataStore.AutoCreateOption
			Get
				Return AutoCreateOption
			End Get
		End Property

		Private Function IDataStore_ModifyData(ParamArray ByVal dmlStatements() As ModificationStatement) As ModificationResult Implements IDataStore.ModifyData
			Return ModifyData(dmlStatements)
		End Function

		Private Function IDataStore_SelectData(ParamArray ByVal selects() As SelectStatement) As SelectedData Implements IDataStore.SelectData
			Return SelectData(selects)
		End Function

		Private Function IDataStore_UpdateSchema(ByVal dontCreateIfFirstTableNotExist As Boolean, ParamArray ByVal tables() As DBTable) As UpdateSchemaResult Implements IDataStore.UpdateSchema
			Return UpdateSchema(dontCreateIfFirstTableNotExist, tables)
		End Function

		#End Region

		#Region "ICommandChannel Members"

		Private Function ICommandChannel_Do(ByVal command As String, ByVal args As Object) As Object Implements ICommandChannel.Do
			Return [Do](command, args)
		End Function

		#End Region
	End Class

	<Serializable> _
	Public Class StoredProcedureArgument
		Public Sub New()
		End Sub

		Public Sub New(ByVal sprocName As Object, ByVal args As Object)
			StoredProcedureName = sprocName.ToString()
			Arguments = CType(args, OperandValue())
		End Sub

		Private fStoredProcedureName As String
		Public Property StoredProcedureName() As String
			Get
				Return fStoredProcedureName
			End Get
			Set(ByVal value As String)
				fStoredProcedureName = value
			End Set
		End Property

		Private fArguments() As OperandValue
		Public Property Arguments() As OperandValue()
			Get
				Return fArguments
			End Get
			Set(ByVal value As OperandValue())
				fArguments = value
			End Set
		End Property
	End Class
End Namespace

File: CS/CommanChannelServiceDataStore/CommandChannelServiceDataStore.cs
using System;
using DevExpress.Xpo.DB;
using System.Web.Services;
using DevExpress.Xpo.Helpers;
using System.Web.Services.Protocols;
using System.Xml.Serialization;
using DevExpress.Data.Filtering;
using System.Collections.Generic;

namespace DXSample.DataStore {
    [WebServiceBinding(Namespace="http://devexpress.example/")]
    public class CommandChannelServiceDataStore :SoapHttpClientProtocol, IDataStore, ICommandChannel {
        public CommandChannelServiceDataStore(string url) : base() {
            Url = url;
        }

        protected virtual AutoCreateOption AutoCreateOption { get { return AutoCreateOption.SchemaAlreadyExists; } }

        [SoapDocumentMethod("http://devexpress.example/ModifyData", RequestNamespace = "http://devexpress.example/", 
            ResponseNamespace = "http://devexpress.example/")]
        public virtual ModificationResult ModifyData(params ModificationStatement[] dmlStatements) {
            return (ModificationResult)Invoke("ModifyData", new object[] { dmlStatements })[0];
        }

        [SoapDocumentMethod("http://devexpress.example/SelectData", RequestNamespace = "http://devexpress.example/", 
            ResponseNamespace = "http://devexpress.example/")]
        public virtual SelectedData SelectData(params SelectStatement[] selects) {
            return (SelectedData)Invoke("SelectData", new object[] { selects })[0];
        }

        protected virtual UpdateSchemaResult UpdateSchema(bool dontCreateIfFirstTableNotExist, params DBTable[] tables) {
            throw new InvalidOperationException("Schema modification is not allowed");
        }

        [SoapDocumentMethod("http://devexpress.example/Do", RequestNamespace = "http://devexpress.example/", 
            ResponseNamespace = "http://devexpress.example/")]
        [XmlInclude(typeof(StoredProcedureArgument))]
        public virtual object Do(string command, object args) {
            return Invoke("Do", new object[] { command, PrepareArguments(args) })[0];
        }

        object PrepareArguments(object args) {
            object[] arguments = args as object[];
            if (arguments == null) return args;
            return new StoredProcedureArgument(arguments[0], arguments[1]);
        }

        #region IDataStore Members

        AutoCreateOption IDataStore.AutoCreateOption {
            get { return AutoCreateOption; }
        }

        ModificationResult IDataStore.ModifyData(params ModificationStatement[] dmlStatements) {
            return ModifyData(dmlStatements);
        }

        SelectedData IDataStore.SelectData(params SelectStatement[] selects) {
            return SelectData(selects);
        }

        UpdateSchemaResult IDataStore.UpdateSchema(bool dontCreateIfFirstTableNotExist, params DBTable[] tables) {
            return UpdateSchema(dontCreateIfFirstTableNotExist, tables);
        }

        #endregion

        #region ICommandChannel Members

        object ICommandChannel.Do(string command, object args) {
            return Do(command, args);
        }

        #endregion
    }

    [Serializable]
    public class StoredProcedureArgument {
        public StoredProcedureArgument() { }

        public StoredProcedureArgument(object sprocName, object args) {
            StoredProcedureName = sprocName.ToString();
            Arguments = (OperandValue[])args;
        }

        private string fStoredProcedureName;
        public string StoredProcedureName {
            get { return fStoredProcedureName; }
            set { fStoredProcedureName = value; }
        }

        private OperandValue[] fArguments;
        public OperandValue[] Arguments {
            get { return fArguments; }
            set { fArguments = value; }
        }
    }
}
